### ISSUE TITLE
Means of targeting minion in salt cli and states and pillar top.sls

### ISSUE THREAD

**Issue by mallniya** [saltstack/salt#39457]
### Description of Issue/Question

Hello, I use saltstack to manage 150+ hosts with different purpose and the problem which always was there is targeting. What I exactly means is following: I see no consistent, centralized and convenient way of grouping my hosts. To the best of my knowledge pillars and grains does not work for this.

Example: we have top.sls both for states and pillar, then it's really nice to have an entity, which we can use for targeting in both tops, like this:
```
{{ saltenv }}:
  - 'G@roles:front':
    - match: compound
    - nginx
```
Pillars doesn't fit this purpose because pillars itself cannot be used in pillar/top.sls.
Currently I use grains for targeting as in example above and I constantly facing annoying aspect of using grains: grains need to be synced with the minion first. And it's not like I complain -- because that's not what grains were intended for. But if minion was restarted or target minion is just new -- there is no way to sync grains with it by role (or group name, whatever unit of grouping is called in your case): only `salt '*' saltutil.sync_all` or `salt 'minion_id' saltutil.sync_all` will do the trick (and waste your time with 150+ hosts), only after this we can use something like `salt -C 'G@roles:front' state.apply`.

One can tell: why not use hardcoded grains in minion config? And the answer is: when amount of host drop to 3 -- I will consider hardcoding, now I need centralized solution (and what does not need to sync everytime I change something).
Another will ask why not use smart named hostnames/minion_ids, which, for example, will contain group/role/env etc. ? Because in real world hostnames naming is not always consistent, and using of hardcoded minion_id on minion side as I mentioned above is a bad idea for wide distributed systems. Targeting by an ID is the thing, that I trying to avoid here (150+ hosts to remember), compare this two options, for instance:
```salt -C 'G@roles:consul and G@env:stage' state.apply consul```
```salt -C 'ns* and front* and ...' state.apply consul``` where compound line will contain 20+ masks of hosts and individual minions, because consul state is needed by many, but not all minions (in which case '*' will work).

So what I'm looking for is third entity of targeting -- not pillar or grain, but something like second pillar namespace called group or role, which indicate what group minion belongs to. Moreover there should be only one file, where minions mapped to their group and ability to join groups, like this:
```
role1:
  - id1

role2:
  -  id2

role3: role1 | role2 |
  - id3
```
because with a lot of minions repeating yourself is a pain and a chance to forget/lost minions.
And of course this entity must be available in states/pillars and salt cli.

**Comment by Ch3LL** [comment]
What about nodegroups?

**Comment by mallniya** [comment]
nodegroups looks like a step in a right direction, but:

1. citing https://docs.saltstack.com/en/latest/topics/targeting/nodegroups.html:

> When adding or modifying nodegroups to a master configuration file, the master must be restarted for those changes to be fully recognized.

NOT a good idea for 150+ master. Moreover:

> Also, if you make changes to your nodegroups, you might need to run salt '*' saltutil.refresh_pillar after restarting the master.

So it's time consuming and could interrupt anyone else on the same master (other admin or Jenkins).

2. referencing nodegroups as follows:

> salt['pillar.get']('master:nodegroups:webserver', [])

means you'll never get to nodegroups in pillars (except for mapping in pillar top.sls).

3. no options for environment-aware nodegroups: what if, for example, I want to join one more minion to group for testing purposes, but I don't want others (anyone who can apply states on the same master to the nodegroup I join minion to) to apply states to this minion?

On the other hand groups looks consistent and centralized enough. And feature mentioned as:

> Nodegroups can reference other nodegroups

just awesome.

**Comment by Ch3LL** [comment]
@mallniya okay so it seems there is nothing that fits your current use case. Currently the top system is pluggable and you are welcome to write your own custom top system to possibly get closer to your goal, but unfortunately that doesn't capture your entire use case because right now we do not have a pluggable targeting system.  To note there is some discussion at the end of this issue https://github.com/saltstack/salt/issues/32723 concerning a pluggable targeting system.

**Comment by Ch3LL** [comment]
Forgot to mention the docs for the master tops system: https://docs.saltstack.com/en/latest/topics/master_tops/

**Comment by mallniya** [comment]
Ok, thanks @Ch3LL . Bad news for me: weak python abilities of mine left no chances to make custom features on my own.
Could this be a feature request? And by "this" I mean (citing myself) "third entity of targeting" from first message.

**Comment by whiteinge** [comment]
@mallniya I think we can make your "second pillar namespace" idea work by generating a Pillar file that contains role information for individual minions. Your above example is pretty similar to the example in the [match.search_by](https://docs.saltstack.com/en/develop/ref/modules/all/salt.modules.match.html#salt.modules.match.search_by) documentation. This function ships with newer versions of Salt but will run on any current version of Salt as a custom module (which is very easy to add; we can talk you through that).

That would allow you to target minions from the CLI via `salt -I roles:role1 test.ping` and would be available to states as `{% set roles = salt.pillar.get('roles'), []) %}`.

Does that look like it will work?

**Comment by mallniya** [comment]
@whiteinge thanks for an option.
In past we used something similar: we didn't use any modules, but just had pillar file for ANY minion with `roles: ...` in it. We abandon this method for role in grains, because with grains we can match roles with pillars in pillar/top.sls. And that's why I have a couple of questions:
1) am I right that roles in this case still wouldn't be available in pillars?
2) could I somehow use roles matching in pillar/top.sls? Like this:
```
{{ saltenv }}:
  - 'I@roles:front': (or other form)
    - match: compound
    - nginx
```
because as I understand this module will not help me with this and large part of my problem is to match minion by role both in pillar and states tops. Can jinja import help me with this?

**Comment by whiteinge** [comment]
Yes! Putting roles in a separate file and importing via Jinja is a great way to side-step the "no pillar from pillar" problem. E.g.:

```
# /srv/pillar/roles.jinja
{% load_yaml as role_defs %}
role1:
  - id1

role2:
  -  id2

role3: role1 | role2 |
  - id3
{% endload %}
{% set roles = salt.match.search_by(role_defs) %}
```

Be sure to expose that computed roles list to the main Pillar generation:

```
# /srv/pillar/roles.sls
{% from "roles.jinja" import roles with context %}
roles: {{ roles | yaml() }}
```

But then you can also import that from any SLS file you want the `roles` list available:

```
# /srv/pillar/myfile.sls
{% from "roles.jinja" import roles with context %}

{% if 'role1' in roles %}
...
{% endif %}
```

**Comment by mallniya** [comment]
ok, looks hopeful, but (sorry) I can't just get grip on this -- let's imagine there are 3 roles and 4 minions (2 fronts and 2 backs):
```
# /srv/pillar/roles.jinja
{% load_yaml as role_defs %}
front:
  - front00.tld
  - front10.tld
uwsgi:
  - back00.tld
  - back10.tld
ldap:
  - front00.tld
  - front10.tld
  - back00.tld
  - back10.tld
{% endload %}
{% set roles = salt.match.search_by(role_defs) %}
```
And we have `/srv/pillar/roles.sls` as you propose. Question -- will `pillar/top.sls` look like this:
```
{{ saltenv }}:
  - '*':
    - roles
```
?
And if yes how it's possible to match exact pillars (I mean pillars other than roles), say, with front or uwsgi role minions in top.sls?
Another question is: could we reuse roles/groups in other roles/groups? In `/srv/pillar/roles.jinja` above, for instance, could we reference to front and uwsgi groups in ldap role instead of repeating minion names, like this:
```
ldap:
  - 'somehow reference front role minions'
  - 'somehow reference uwsgi role minions'
```
?
